Release v1.4.0 (#5363)
grafana/k6: release notes/v1.4.0.md

k6 `v1.4.0` is here 🎉! This release includes:

- Changes in the Browser module:
  - `page.waitForRequest` for waiting on specific HTTP requests.
  - `QueryAll` methods now return elements in DOM order.
  - `locator.evaluate` and `locator.evaluateHandle` for executing JavaScript code in the page context with access to the matching element.
  - `page.unroute(url)` and `page.unrouteAll` for removing routes registered with `page.route`.
- Other notable changes:
  - OpenTelemetry output graduated from experimental to stable status.

## Breaking changes

As per our [stability guarantees](https://grafana.com/docs/k6/latest/reference/versioning-and-stability-guarantees/), breaking changes across minor releases are allowed only for experimental features.

### Breaking changes for experimental modules

- [#5164](https://github.com/grafana/k6/pull/5164) OpenTelemetry output now exports rate metrics as a single counter with `zero`/`nonzero` labels instead of separate metrics.
- [#5333](https://github.com/grafana/k6/pull/5333) OpenTelemetry output configuration: `K6_OTEL_EXPORTER_TYPE` is deprecated in favor of `K6_OTEL_EXPORTER_PROTOCOL` to align with OpenTelemetry standards.

> - [#5320](https://github.com/grafana/k6/pull/5320), [#5239](https://github.com/grafana/k6/pull/5239), [#5342](https://github.com/grafana/k6/pull/5342) Automatic extension resolution now only inspects ES module `import` statements and no longer supports CommonJS `require()` calls. CommonJS `require()` calls are dynamic and it is not possible to know for certain if they will be called, or if they will be called with static strings - the only way they were even previously loaded. This functionality was a quirk of the previous implementation and had numberous problems. Additionally, `use k6` directives are now only recognized when they appear at the beginning of files (after optional shebang and whitespace/comments). This was the initial intention, but due to implementation bugs it was not actually was happening.

Reviewer: I would still want to have it in breaking changes as a thing that has changed, even if it might have not been intentional to work like this.

> - [#5320](https://github.com/grafana/k6/pull/5320), [#5239](https://github.com/grafana/k6/pull/5239), [#5342](https://github.com/grafana/k6/pull/5342) Automatic extension resolution now only inspects ES module `import` statements and no longer supports CommonJS `require()` calls. CommonJS `require()` calls are dynamic and it is not possible to know for certain if they will be called, or if they will be called with static strings - the only way they were even previously loaded. This functionality was a quirk of the previous implementation and had numberous problems. Additionally, `use k6` directives are now only recognized when they appear at the beginning of files (after optional shebang and whitespace/comments). This was the initial intention, but due to implementation bugs it was not actually was happening.

Reviewer 2: Then you should move it out from the experimental modules list, as this is not the case.

k6 `v1.4.0` is here 🎉! This release includes:

- Changes in the Browser module:
  - `page.waitForRequest` for waiting on specific HTTP requests.
  - `QueryAll` methods now return elements in DOM order.
  - `locator.evaluate` and `locator.evaluateHandle` for executing JavaScript code in the page context with access to the matching element.
  - `page.unroute(url)` and `page.unrouteAll` for removing routes registered with `page.route`.
- Other notable changes:

>   - OpenTelemetry output graduated from experimental to stable status.

Reviewer 3: Not a strong opinion, but should we list the last item first, rather than having a separate list?

```suggestion
- OpenTelemetry output graduated from experimental to stable status.
- Changes in the Browser module:
  - `page.waitForRequest` for waiting on specific HTTP requests.
  - `QueryAll` methods now return elements in DOM order.
  - `locator.evaluate` and `locator.evaluateHandle` for executing JavaScript code in the page context with access to the matching element.
  - `page.unroute(url)` and `page.unrouteAll` for removing routes registered with `page.route`.
```

```

### Automatic extension resolution improvements [#5320](https://github.com/grafana/k6/pull/5320), [#5239](https://github.com/grafana/k6/pull/5239), [#5342](https://github.com/grafana/k6/pull/5342), [#5332](https://github.com/grafana/k6/pull/5332), [#5240](https://github.com/grafana/k6/pull/5240)

Automatic extension resolution has been completely reimplemented to use k6's internal module loader instead of the external `k6deps`/esbuild pipeline. This change brings significant improvements in reliability and maintainability.

As part of the rewrite a few issues and unintended "features' were found, namely:
1. trying to follow `require` calls, which due to their dynamic nature doesn't work particularly stable. That is depending on where and how `require` call was used it was possible that k6 might decided it is needed or not. And it definitely doesn't work wiht actually using string variables. Support for CommonJS is mostly for backwards compatibiliy, so after internal discussion we opted to not support this at all. We could bring back this until v2, if there is enough interest. But long term it is inteded for this to not be part of k6.

> 2. "use k6 with ..." directives, were parsed from the whole file instead of just the beginning, which leads to numerous problems, and was not the intented case. As such they are now only parsed at the beginning of files (not only the main one) with potential empty lines and comments infront.

Reviewer 3: ```suggestion
1. Trying to follow `require` calls, which, due to their dynamic nature, don't work particularly stably. That is, depending on where and how the `require` call was used, k6 might decide whether it is needed or not. And it definitely doesn't work when using actual string variables. Support for CommonJS is primarily for backward compatibility, so after an internal discussion, we opted not to support it at all. We could bring this back until v2, if there is enough interest. However, in the long term, it is intended that this not be part of k6.
2. "use k6 with ..." directives were parsed from the whole file instead of just the beginning, which leads to numerous problems, and was not the intended case. As such, they are now only parsed at the beginning of files (not just the main one) with potential empty lines and comments preceding them.
```

**Example:**

```javascript
// main.js
"use k6 with k6/x/faker"
import { faker } from 'k6/x/faker';
import { helper } from './utils.js';

export default function() {
  console.log(faker.name());
  helper();
}
```

> Or an axample using the directive but commonJS

Reviewer 3: ```suggestion
Or, an example using the directive, but using commonJS
```

- [#5216](https://github.com/grafana/k6/pull/5216) Refactors `TestPageOnResponse` to fail fast on errors. Thank you, @rMaxiQp for the contribution.
- [#5203](https://github.com/grafana/k6/pull/5203), [#5201](https://github.com/grafana/k6/pull/5201) Adds WinGet support to the project.
- [#5220](https://github.com/grafana/k6/pull/5220) Switches from gouuid to google/uuid. Thank you, @mikelolasagasti for the contribution.
- [#5344](https://github.com/grafana/k6/pull/5344), [#5345](https://github.com/grafana/k6/pull/5345) Uses .env context for ghcr login when publishing docker images.
- [#5331](https://github.com/grafana/k6/pull/5331) Reports `require` calls during the initial loading of modules, instead for each call in all execution of all VUs.
- [#5218](https://github.com/grafana/k6/pull/5218) Adds user AgnesToulet to the auto assign issue workflow.

## _Optional_ Roadmap

> _Discussion of future plans_

Author: Thanks for the reminder, Removed
